File: 9-1.js
const example = `R 4
U 4
L 3
D 1
R 4
D 1
L 5
R 2`;

require("fs").readFile("9.txt", "utf-8", (err, data) => {
    const steps = data
        .split("\n")
        .filter(Boolean)
        .map(l => l.split(" "))
        .reduce((steps, [direction, amount]) => {
            while (amount--) steps.push(direction);
            return steps;
        }, [])
        .reduce(
            (heads, s) => {
                const { x, y } = heads[heads.length - 1];
                heads.push({
                    x: s === "L" ? x - 1 : s === "R" ? x + 1 : x,
                    y: s === "D" ? y - 1 : s === "U" ? y + 1 : y,
                });
                return heads;
            },
            [{ x: 0, y: 0 }],
        )
        .reduce(
            (tails, head) => {
                let { x, y } = tails[tails.length - 1];

                const dx = head.x - x;
                const dy = head.y - y;
                if (Math.abs(dx) > 1 || Math.abs(dy) > 1) {
                    x = dx === 0 ? x : dx > 0 ? x + 1 : x - 1;
                    y = dy === 0 ? y : dy > 0 ? y + 1 : y - 1;
                }

                tails.push({ x, y });
                return tails;
            },
            [{ x: 0, y: 0 }],
        )
        .reduce(
            (positions, tail) => {
                positions.add(`${tail.x},${tail.y}`);
                return positions;
            },
            new Set(),
        );

    console.log(steps.size);
});

File: 9-1.c
#include "prelude.h"

static const char *example = "R 4\n"
			     "U 4\n"
			     "L 3\n"
			     "D 1\n"
			     "R 4\n"
			     "D 1\n"
			     "L 5\n"
			     "R 2\n";

typedef struct {
	int x, y;
} Pos;

typedef struct Path Path;
struct Path {
	Pos pos;
	Path *next;
};


static void follow(const Pos *head, Pos *tail) {
	int dx = head->x - tail->x;
	int dy = head->y - tail->y;

	if (abs(dx) > 1 || abs(dy) > 1) {
		if (dx > 0) tail->x++;
		if (dx < 0) tail->x--;
		if (dy > 0) tail->y++;
		if (dy < 0) tail->y--;
	}
}

static void move(Pos *head, char direction) {
	switch (direction) {
	case 'U': head->y--; break;
	case 'D': head->y++; break;
	case 'L': head->x--; break;
	case 'R': head->x++; break;
	}
}

static void destroy(Path *path) {
	if (!path) return;
	destroy(path->next);
	free(path);
}

static void push(Path *path, Pos pos) {
	assert(path);
	if (path->pos.x == pos.x && path->pos.y == pos.y) return;
	while (path->next) {
		path = path->next;
		if (path->pos.x == pos.x && path->pos.y == pos.y) return;
	}
	path->next = malloc(sizeof(Path));
	path->next->pos = pos;
}

static int len(const Path *path) {
	assert(path);
	int n = 1;

	while ((path = path->next))
		n++;

	return n;
}

static int solve(char *input) {
	Pos head = {.x = 0, .y = 0};
	Pos tail = {.x = 0, .y = 0};
	Path *path = malloc(sizeof(Path));
	path->pos = tail;
	path->next = NULL;

	LINES(line, input) {
		if (!strcmp(line, "")) continue;
		char direction;
		int amount;
		assert(sscanf(line, "%c %d", &direction, &amount) == 2);
		while (amount--) {
			move(&head, direction);
			follow(&head, &tail);
			push(path, tail);
		}
	}


	int count = len(path);

	destroy(path);

	return count;
}

int main() {
	RUN(solve, strdup(example));
	RUN(solve, read_to_string("9.txt"));
}
